# <center>New Program Creation: A Case Study</center>
## <center> Undergraduate Software Engineering Curricula at the Univeristy of Arizona </center>

<center>
    <b>Gregory L. Heileman$^\dagger$</b> <br>
    $^\dagger$Department of Electrical & Computer Engineering <br>
    University of Arizona <br>
    heileman@arizona.edu
</center>

## 1  Introduction
In this case study we consider the creation of a new undergraduate software engineering program at the University of Arizona. This will include consideration of the expected demand for the major as well as the employment opportunites that will be available to the graduates of this programs. Next we will consider similar programs at peer institions, which will include an analysis of the curricula offered by Software Engineering programs at these institutions. 

## 2  Market Demand

Software developer is ranked as the best technology job in America by the U.S. News & World Report

Bureau of Labor Statistics predicts a 30% growth in software engineering jobs from 2010 through 2020 (1.85M to 2.4M)

Co-ops/internships available at companies such as Lockheed Martin, Rockwell Collins and Citigroup

<Complete using Stephanie's data>



## 3  ABET Requirements 

In this section we consider the manner in which software engineering progrmas are offered at a number of peer institutions. Because our goal is to create a software engineering program that is accredited by ABET, we will consider only those peer programs that seek similar ABET accreditation. 

The requirements associated with ABET accreditation that are relevant to new program creation include the following items.  <Describe EAC and CSAB>

**Program Educational Objectives.** In order for a program to recieve ABET accredition it must identify a set of *program education objectives* (PEOs) that broadly state what the program's graduates are expected to attain within a few years after graduation. These PEOs must be created and maintained so as to address the needs of the program's constituents. Thus, in constructing the PEOs for the University of Arizona software engineering program, we must indentify constituents and how the program will serve their needs. 

**Student Outcomes.** Using ABET terminology, student outcomes are program-level learning outcomes that the curriculum associted with *any* ABET-accreditied program must support. By attaining these outcomes, students will be prepared to enter the practice of engineering.  These outcomes include:

1. an ability to identify, formulate, and solve complex engineering problems by applying principles of engineering, science, and mathematics
2. an ability to apply engineering design to produce solutions that meet specified needs with consideration of public health, safety, and welfare, as well as global, cultural, social, environmental, and economic factors
3. an ability to communicate effectively with a range of audiences
4. an ability to recognize ethical and professional responsibilities in engineering situations and make informed judgments, which must consider the impact of engineering solutions in global, economic, environmental, and societal contexts
5. an ability to function effectively on a team whose members together provide leadership, create a collaborative and inclusive environment, establish goals, plan tasks, and meet objectives
6. an ability to develop and conduct appropriate experimentation, analyze and interpret data, and use engineering judgment to draw conclusions
7. an ability to acquire and apply new knowledge as needed, using appropriate learning strategies.

In addition, ABET 
    These program criteria apply to engineering programs that include “software” or similar modifiers in their titles.
1. Curriculum.
The curriculum must provide both breadth and depth across the range of engineering and computer science topics implied by the title and objectives of the program.
The curriculum must include computing fundamentals, software design and construction, requirements analysis, security, verification, and validation; software engineering processes and tools appropriate for the development of complex software systems; and discrete mathematics, probability, and statistics, with applications appropriate to software engineering.
2. Faculty.
The program must demonstrate that faculty members teaching core software engineering topics have an understanding of professional practice in software engineering and maintain currency in their areas of professional or scholarly specialization.

## 4  Peer Comparisons
There are a total of 37 ABET-accredited bachelor's level programs in Software Engineering, with 30 of these located in the United States (see: https://www.abet.org/accreditation/find-programs/).

Note: The UC-Irvine program is NOT listed as having ABET accreditation, and Carnegie Mellon offers a Software Engineerig Minor that is not ABET accredited.

In order to conduct curricular analyses of peer schools, we first need to load the packages associated with the Curricular Analtyics Toolbox.

In [159]:
using CurricularAnalytics, CurricularVisualization
using DataFrames
using GLM

***Arizona State University.*** This program is located in the Schools of Engineering at ASU and leads to the Bachelor of Science in software engineering degree. They offer a 3-year accelerated programs, as well as an accelerated 4+1 program leading to a Master's of Science in software engineering.

In [160]:
ASU_plan = read_csv("./programs/Arizona-State-BSSE.csv")
visualize(ASU_plan, notebook=true, scale=0.9)

WebIO.Scope(WebIO.Node{WebIO.DOM}(WebIO.DOM(:html, :iframe), Any[""], Dict{Symbol,Any}(:src => "http://localhost:8156",:id => "curriculum",:events => Dict{Symbol,WebIO.JSString}(:load => WebIO.JSString("(function (){this.contentWindow.postMessage({\"options\":{\"hideTerms\":false,\"edit\":false},\"curriculum\":{\"name\":\"Software Engineering\",\"dp_name\":\"2020-21 Degree Plan\",\"curriculum_terms\":[{\"name\":\"Term 1\",\"curriculum_items\":[{\"curriculum_requisites\":[],\"name\":\"CSE 110\",\"metrics\":{\"centrality\":0,\"complexity\":23.0,\"blocking factor\":16,\"delay factor\":7.0},\"nameSub\":\"Principles of Programming\",\"id\":1,\"credits\":3,\"nameCanonical\":\"Programming I\"},{\"curriculum_requisites\":[],\"name\":\"MAT 265\",\"metrics\":{\"centrality\":0,\"complexity\":26.0,\"blocking factor\":19,\"delay factor\":7.0},\"nameSub\":\"Calculus for Engineers I\",\"id\":2,\"credits\":3,\"nameCanonical\":\"Calculus I\"},{\"curriculum_requisites\":[],\"name\":\"ASU 101-CSE\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"The ASU Experience\",\"id\":3,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"ENG 101\",\"metrics\":{\"centrality\":0,\"complexity\":6.0,\"blocking factor\":3,\"delay factor\":3.0},\"nameSub\":\"First-year Composition\",\"id\":4,\"credits\":3,\"nameCanonical\":\"First-year Composition I\"},{\"curriculum_requisites\":[],\"name\":\"FSE 100\",\"metrics\":{\"centrality\":0,\"complexity\":6.0,\"blocking factor\":3,\"delay factor\":3.0},\"nameSub\":\"Introduction to Engineering \",\"id\":5,\"credits\":2,\"nameCanonical\":\"Introduction to Engineering\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Social-Behavioral Sciences\",\"id\":6,\"credits\":3,\"nameCanonical\":\"\"}],\"id\":1},{\"name\":\"Term 2\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":1,\"target_id\":7,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"CSE 205\",\"metrics\":{\"centrality\":72,\"complexity\":20.0,\"blocking factor\":13,\"delay factor\":7.0},\"nameSub\":\"Object-Oriented Programming and Data Structures\",\"id\":7,\"credits\":3,\"nameCanonical\":\"Data Structures\"},{\"curriculum_requisites\":[{\"source_id\":4,\"target_id\":8,\"type\":\"CurriculumCorequisite\"}],\"name\":\"EGR 104\",\"metrics\":{\"centrality\":0,\"complexity\":2.0,\"blocking factor\":0,\"delay factor\":2.0},\"nameSub\":\"Critical Inquiry in Engineering\",\"id\":8,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":4,\"target_id\":9,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"ENG 102\",\"metrics\":{\"centrality\":3,\"complexity\":4.0,\"blocking factor\":1,\"delay factor\":3.0},\"nameSub\":\"First-year Composition\",\"id\":9,\"credits\":3,\"nameCanonical\":\"First-year Composition II\"},{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":10,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MAT 266\",\"metrics\":{\"centrality\":9,\"complexity\":6.0,\"blocking factor\":3,\"delay factor\":3.0},\"nameSub\":\"Calculus for Engineers II\",\"id\":10,\"credits\":3,\"nameCanonical\":\"Calculus II\"},{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Humanities, Arts and Design AND Cultural Diversity in the U.S.\",\"id\":11,\"credits\":3,\"nameCanonical\":\"\"}],\"id\":2},{\"name\":\"Term 3\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":12,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MAT 243\",\"metrics\":{\"centrality\":69,\"complexity\":19.0,\"blocking factor\":12,\"delay factor\":7.0},\"nameSub\":\"Discrete Mathematical Structures\",\"id\":12,\"credits\":3,\"nameCanonical\":\"Discrete Mathematics\"},{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":13,\"type\":\"CurriculumPrerequ

This program was first ABET-accredited in 2014, and it currently has the following program educational objectives.

Within a few years of completion, graduates exhibit the following characteristics:
- Be agile software developers with a comprehensive set of skills appropriate to the needs of the dynamic global computing-based society.
- Capable of team and organizational leadership in computing project settings, and have a broad understanding of ethical application of computing-based solutions to societal and organizational problems.
- Acquire skills and knowledge to advance their career, including continually upgrading professional, communication, analytic, and technical skills.

According to the ASU website, in 2018 there werew 1,268	students enrolled in this program, with 122 degrees awarded,  and in 2019, there werew 1,322 student enrolled in this program, with 143 degrees awarded.

The prefix for software engineering courses at ASU is "SER."  Note in the degree plan above that the first SER course does not occur until the first semester of the sophomore year.  There total number of software engineering courses (i.e., courses with SER prefixes) in the ASU curriculum is:

In [161]:
SSE_ASU = mapreduce(x->x.prefix == "SER", +, ASU_plan.curriculum.courses)

16

That is, in order to support this curriculum, ASU was required to create at least 16 new courses in the software engineering area.  The sofware engineering courses account for the following percentage of the entire curriculum:

In [162]:
round(SSE_ASU/length(ASU_plan.curriculum.courses) * 100)

39.0

***Stevens Institute of Technolgy.*** This program is offered jointly by the School of Systems and Enterprises and the School of Engineering and Science, and leads to the Bachelor of Engineering in software engineering. The first graduates from this program occured in 2018. They also offer an accelerated Master's program. A minor in Software Engineering requires seven courses, two of which are required in any ABET-accredited computer science or computer engineering program. Thus, taking the other four courses as electives, it seems that it would be fairly easy for computer science of computer engineering undergraduate students at Stevens to earn a Software Engineering Minor.

The program has a high number of total credit hours, but the goal of the program is to provide depth in both software engineering and systems engineering.

In [163]:
SIT_plan = read_csv("./programs/Stevens-Inst-BESE.csv")
visualize(SIT_plan, notebook=true)

WebIO.Scope(WebIO.Node{WebIO.DOM}(WebIO.DOM(:html, :iframe), Any[""], Dict{Symbol,Any}(:src => "http://localhost:8156",:id => "curriculum",:events => Dict{Symbol,WebIO.JSString}(:load => WebIO.JSString("(function (){this.contentWindow.postMessage({\"options\":{\"hideTerms\":false,\"edit\":false},\"curriculum\":{\"name\":\"Software Engineering\",\"dp_name\":\"2020-21 Degree Plan\",\"curriculum_terms\":[{\"name\":\"Term 1\",\"curriculum_items\":[{\"curriculum_requisites\":[{\"source_id\":2,\"target_id\":1,\"type\":\"CurriculumStrictCorequisite\"}],\"name\":\"CH 115\",\"metrics\":{\"centrality\":7,\"complexity\":7.0,\"blocking factor\":3,\"delay factor\":4.0},\"nameSub\":\"General Chemistry I \",\"id\":1,\"credits\":3,\"nameCanonical\":\"Chemistry I\"},{\"curriculum_requisites\":[],\"name\":\"CH 117\",\"metrics\":{\"centrality\":0,\"complexity\":8.0,\"blocking factor\":4,\"delay factor\":4.0},\"nameSub\":\"General Chemistry Laboratory I \",\"id\":2,\"credits\":1,\"nameCanonical\":\"Chemistry I Lab\"},{\"curriculum_requisites\":[],\"name\":\"E 101\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Engineering Experience I \",\"id\":3,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"E 120\",\"metrics\":{\"centrality\":0,\"complexity\":11.0,\"blocking factor\":5,\"delay factor\":6.0},\"nameSub\":\"Engineering Graphics \",\"id\":4,\"credits\":1,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":4,\"target_id\":5,\"type\":\"CurriculumCorequisite\"},{\"source_id\":6,\"target_id\":5,\"type\":\"CurriculumCorequisite\"}],\"name\":\"E 121\",\"metrics\":{\"centrality\":36,\"complexity\":10.0,\"blocking factor\":4,\"delay factor\":6.0},\"nameSub\":\"Engineering Design I: Introduction to Systems Thinking \",\"id\":5,\"credits\":2,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"E 115\",\"metrics\":{\"centrality\":0,\"complexity\":15.0,\"blocking factor\":9,\"delay factor\":6.0},\"nameSub\":\"Introduction to Programming \",\"id\":6,\"credits\":2,\"nameCanonical\":\"Programming I\"},{\"curriculum_requisites\":[],\"name\":\"CAL 103\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Writing And Communications Colloquium \",\"id\":7,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"MA  121;122\",\"metrics\":{\"centrality\":0,\"complexity\":18.0,\"blocking factor\":12,\"delay factor\":6.0},\"nameSub\":\"Differential Calculus; Integral Calculus \",\"id\":8,\"credits\":4,\"nameCanonical\":\"\"}],\"id\":1},{\"name\":\"Term 2\",\"curriculum_items\":[{\"curriculum_requisites\":[],\"name\":\" \",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"Science Elective\",\"id\":9,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":8,\"target_id\":10,\"type\":\"CurriculumCorequisite\"}],\"name\":\"PEP 111\",\"metrics\":{\"centrality\":20,\"complexity\":13.0,\"blocking factor\":7,\"delay factor\":6.0},\"nameSub\":\"Mechanics\",\"id\":10,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":5,\"target_id\":11,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"E 122\",\"metrics\":{\"centrality\":30,\"complexity\":9.0,\"blocking factor\":3,\"delay factor\":6.0},\"nameSub\":\"Engineering Design II: Field Sustainable Systems with Sensors \",\"id\":11,\"credits\":2,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[],\"name\":\"CAL 105\",\"metrics\":{\"centrality\":0,\"complexity\":1.0,\"blocking factor\":0,\"delay factor\":1.0},\"nameSub\":\"CAL Colloquium: Knowledge, Nature, Culture\",\"id\":12,\"credits\":3,\"nameCanonical\":\"\"},{\"curriculum_requisites\":[{\"source_id\":8,\"target_id\":13,\"type\":\"CurriculumPrerequisite\"}],\"name\":\"MA  123;124\",\"metrics\":{\"centrality\":12,\"complexity\":11.0,\"blocking factor\":6,\"delay factor\":5.0},\"nameS

Even though this programs claims they are ABET-accredited, ABET does *not* list them as currently having an accredited program in software engineering.

The Stevens website mentions that graduates from this program will be able to:
- Design and develop software components for complex systems.
- Communicate with engineers of all disciplines in the languages and methods of those engineers.
- Understand the overall systems context for their projects and apply systems thinking in designing solutions that integrate components of different types, such as hardware, software, and people.

Note: in my view, these do *not* read like ABET PEOs.  I also find it problematic that I cannot locate the ABET-requried enrollment and graduation data on the program's website.  I have asked the program for this information, but they have not provided it to me.

The prefix for software engineering courses at Stevens is "SSW."  Note in the degree plan above that the first SER course does not occur until the second semester of the sophomore year.  The total number of software engineering courses (i.e., courses with SSW prefixes) in the curriculum is:

In [164]:
SSW_SIT = mapreduce(x->x.prefix == "SSW", +, SIT_plan.curriculum.courses)

10

Specifically, in order to support this curriculum, Stevens was required to create only 10 new courses in the software engineering area. Thus, even though this program has far more credit hours than the ASU software engineering program, it also involves 30% fewer software engineering courses.

The Stevens sofware engineering courses account for the following percentage of the entire curriculum:

In [165]:
round(SSW_SIT/length(SIT_plan.curriculum.courses) * 100)

20.0

### 4.1 Projected Program Enrollment

The follwoing enrollment data, which a program must provide to the public as a part of their ABET accreditions, was obtained from the websites of:

 - Arizona State University 
 - Auburn University
 - Mississippi State University
 - Rochester Institute of Technology
 - Rose-Hulman Institute of Technology
 - University of Texas at Arlington
 - University of Texas at Dallas
 - Washington State University
 
Enrollment data was collected for the follwing commonly-offered programs:
 
 - biomedical engineering
 - chemical engineering 
 - civil engineering
 - computer science
 - electrical engineering
 - mechanical engineering
 - software engineering

In [166]:
peer_enrollment_data_2019 = DataFrame(School = ["ASU", "Auburn", "Miss St U", "Rochester", "Rose Hulman", "UT-Arlington", "UT-Dallas ", "Wash St U"], 
    Biomed_Eng = [699, 107, 245, 289, 146, 54, 611, 159],
    Chem_Eng = [648, 610, 413, 240, 180, missing, missing, 204],
    Civil_Eng = [726, 555, 499, 258, 121, 95, missing, 461],
    Comp_Sci = [2981, 646, 408, 984, 330, 127, 2350, 696],
    Elec_Eng = [2369, 399, 380, 564, 151, 62, 619, 281],
    Mech_Eng = [1505, 1275, 979, 770, 607, 128, 1099, 859],
    Soft_Eng = [1322, 388, 159, 551, 70, 23, 480, 141]);

In [167]:
showall(peer_enrollment_data_2019)

8×8 DataFrame
│ Row │ School       │ Biomed_Eng │ Chem_Eng │ Civil_Eng │ Comp_Sci │ Elec_Eng │
│     │ String       │ Int64      │ Int64?   │ Int64?    │ Int64    │ Int64    │
├─────┼──────────────┼────────────┼──────────┼───────────┼──────────┼──────────┤
│ 1   │ ASU          │ 699        │ 648      │ 726       │ 2981     │ 2369     │
│ 2   │ Auburn       │ 107        │ 610      │ 555       │ 646      │ 399      │
│ 3   │ Miss St U    │ 245        │ 413      │ 499       │ 408      │ 380      │
│ 4   │ Rochester    │ 289        │ 240      │ 258       │ 984      │ 564      │
│ 5   │ Rose Hulman  │ 146        │ 180      │ 121       │ 330      │ 151      │
│ 6   │ UT-Arlington │ 54         │ missing  │ 95        │ 127      │ 62       │
│ 7   │ UT-Dallas    │ 611        │ missing  │ missing   │ 2350     │ 619      │
│ 8   │ Wash St U    │ 159        │ 204      │ 461       │ 696      │ 281      │

│ Row │ Mech_Eng │ Soft_Eng │
│     │ Int64    │ Int64    │
├─────┼──────────┼──────────┤
│ 1 

### Linear Regression Model.

Below we fit the enrollment data using a ordinary least squares linear regression. The dependent variable in this model is `software engineering enrollment`, and the independent variables are enrollments in all of the other programs, with the excpetion of chemical engineerig (as a number of the selected schools do not offer that program). Thus, we are trying to determine the relationship between software engineerig enrollments, and the enrollments in other engineering programs offered at the same school. 

In [168]:
ols = fit(LinearModel, 
    @formula(Soft_Eng ~ Biomed_Eng + Civil_Eng + Comp_Sci + Elec_Eng + Mech_Eng),
    peer_enrollment_data_2019)

StatsModels.TableRegressionModel{LinearModel{GLM.LmResp{Array{Float64,1}},GLM.DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}

Soft_Eng ~ 1 + Biomed_Eng + Civil_Eng + Comp_Sci + Elec_Eng + Mech_Eng

Coefficients:
─────────────────────────────────────────────────────────────────────────────
                  Coef.  Std. Error      t  Pr(>|t|)    Lower 95%   Upper 95%
─────────────────────────────────────────────────────────────────────────────
(Intercept)  -34.026     255.643     -0.13    0.9158  -3282.27     3214.22
Biomed_Eng    -0.348615    1.61829   -0.22    0.8649    -20.911      20.2137
Civil_Eng     -0.555588    1.04212   -0.53    0.6882    -13.7969     12.6858
Comp_Sci       0.299436    0.608575   0.49    0.7089     -7.43324     8.03212
Elec_Eng       0.271235    0.869113   0.31    0.8074    -10.7719     11.3144
Mech_Eng       0.327285    0.553263   0.59    0.6599     -6.70259     7.35716
────────────────────────────────────────────────

### Goodness of Fit.
The $R^2$ value of this model provides a measure of how well the model fits the data. In particular, it measures the percentage of the variance in the dependent variable (software engineering enrollment) that is explained by the collection of independent variables (other program enrollments). 

In [169]:
r2(ols)

0.967372753359513

Thus, given that 97% of the variation in software engineering enrollments can be explained by the enrollments in the other programs, this model likely can be used to provide an useful prediction for the eventual enrollment in a new software engineering program at the Univeristy of Arizona.

### Predicted Enrollment.

In [170]:
UA_enrollment_data_2019 = DataFrame(School = ["UA"], 
    Biomed_Eng = [163],
    Chem_Eng = [258],
    Civil_Eng = [123],
    Comp_Sci = [1626],
    Elec_Eng = [390],
    Mech_Eng = [454],
    Soft_Eng = [missing]);

In [171]:
predict(ols, UA_enrollment_data_2019)

1-element Array{Union{Missing, Float64},1}:
 582.0644041096086

Thus, at steady state, we should expect to have roughly 582 students enrolled in a software engineering program at the Univeristy of Arizona.

## 3. Basic Analytics
Now that we have created the curriculum, associated degree plan, and verified there are no errors, we can quantitatively analyze the plan, and more specifically, the courses. 


## Basic Metrics
If more masic metrics are desired such as:

1. Total credit hours
2. Avg. credits per term
3. Number of terms
4. Max. credits in a term
5. Min. credit term
6. Credit hour variance
7. Max. credit term

The following command may be run. Remember that the variable inside of () is whatever you set your degree plan to equal, in our case dp:

In [44]:
basic_metrics(ASU_plan)
ASU_plan.metrics

Dict{String,Any} with 8 entries:
  "total credit hours"         => 114
  "avg. credits per term"      => 14.25
  "min. credits in a term"     => 6
  "term credit hour std. dev." => 3.15238
  "number of terms"            => 8
  "max. credits in a term"     => 16
  "min. credit term"           => 8
  "max. credit term"           => 3

## Conclusions 


## References

Heileman, G. L., Abdallah, C.T., Slim, A., and Hickman, M. (2018). Curricular analytics: A framework for quantifying the impact of curricular reforms and pedagogical innovations. www.arXiv.org, arXiv:1811.09676 [cs.CY].

